Rewrite doins by python (bug 624526).

zmedico opened the pull request:
doins is written in bash. However, specifically in case that
too many files are installed, it is very slow.
This CL rewrites the script in python for performance.

BUG=chromium:712659
TEST=time (./setup_board --forace && \
     ./build_package --withdev && \
     ./build_image --noenable_rootfs_verification test)
===Before===
real    21m35.445s
user    93m40.588s
sys     21m31.224s

===After===
real    17m30.106s
user    94m1.812s
sys     20m13.468s

Change-Id: Ib10f623961ba316753d58397cff5e72fbc343339
Reviewed-on: https://chromium-review.googlesource.com/559225
X-Chromium-Bug: 712659
X-Chromium-Bug-url: https://bugs.chromium.org/p/chromium/issues/detail?id=712659
X-Gentoo-Bug: 624526
X-Gentoo-Bug-url: https://bugs.gentoo.org/624526

mrueg commented on bin/ebuild-helpers/doins:
```diff
@@ -44,116 +44,17 @@ if [[ ${INSDESTTREE#${ED}} != "${INSDESTTREE}" ]]; then
-rm -rf "$TMP"
-[[ $failed -ne 0 || $success -eq 0 ]] && { __helpers_die "${helper} failed"; exit 1; } || exit 0
+"${PORTAGE_PYTHON:-/usr/bin/python}" \
+	"${PORTAGE_BIN_PATH:-/usr/lib/portage/bin}"/doins.py \
```
Is /usr/lib/portage/bin a safe fallback? Shouldn't we check for it being unset instead?

mrueg commented on bin/ebuild-helpers/doins:
```diff
@@ -44,116 +44,17 @@ if [[ ${INSDESTTREE#${ED}} != "${INSDESTTREE}" ]]; then
-mkdir -p "$TMP"/{1,2}
+# Use safe cwd, avoiding unsafe import for bug #469338.
+export __PORTAGE_HELPER_CWD=${PWD}
+cd "${PORTAGE_PYM_PATH:-/usr/lib/portage/pym}"
```
Is this path a safe fallback?

mrueg commented on bin/doins.py:
```diff
@@ -0,0 +1,341 @@
+		"""Installs a file at |source| into |dest_dir| in process."""
+		dest = os.path.join(dest_dir, os.path.basename(source))
+		try:
+			# TODO: Consider to use portage.util.file_copy.copyfile
```
Possible to get this TODO done before merging?

mrueg commented on bin/doins.py:
```diff
@@ -0,0 +1,341 @@
+	def run(self, dest):
+		"""Installs a dir into |dest| by 'install' command."""
+		command = ['install', '-d'] + self._split_options + [dest]
+		subprocess.call(command)
```
Could need some exception handling, if subprocess.call fails

mrueg commented on bin/doins.py:
```diff
@@ -0,0 +1,341 @@
+	# supported.
+	# In Python 3, the prefix of octal int must be '0o' rather than '0'.
+	# So, set base explicitly in that case.
+	return int(mode, 8 if re.search(r'^0[0-7]*$', mode) else 0)
```
Split up into multiple lines for readability and consider compiling the regex

zmedico commented:
Updated with the version merged in [chromiumos/third_party/portage_tool](https://chromium.googlesource.com/chromiumos/third_party/portage_tool/+/dd8ad7251ab85091ebcac11b721eb82e21d82256), and added a commit to optimize with copyfile from portage.util.file_copy.
